### ISSUE TITLE
Understanding mathematical operations on quantities

### ISSUE THREAD

**Issue by RaphaelRobidas** [iliekturtles/uom#482]
I must be misunderstanding how to carry out operations on quantities with units. For example, let's consider this very simple example:
```rust
use uom::si::f32::*;
use uom::si::length::meter;

fn main() {
    let pos: [Length; 3] = [Length::new::<meter>(1.0), Length::new::<meter>(2.0), Length::new::<meter>(3.0)];
    let length: Length = (pos[0].powi(2) + pos[1].powi(2) + pos[2].powi(2)).sqrt();
}
```

The program cannot compile due to errors with the power operations:
```rust
error[E0277]: cannot multiply `PInt<UInt<UTerm, B1>>` by `{integer}`
   --> exa/src/main.rs:7:39
    |
7   |     let length: Length = (pos[0].powi(2) + pos[1].powi(2) + pos[2].powi(2)).sqrt();
    |                                  ---- ^ no implementation for `PInt<UInt<UTerm, B1>> * {integer}`
    |                                  |
    |                                  required by a bound introduced by this call
    |
    = help: the trait `std::ops::Mul<{integer}>` is not implemented for `PInt<UInt<UTerm, B1>>`
    = help: the following other types implement trait `std::ops::Mul<Rhs>`:
              <PInt<U> as std::ops::Mul<Z0>>
              <PInt<U> as std::ops::Mul<ATerm>>
              <PInt<Ul> as std::ops::Mul<PInt<Ur>>>
              <PInt<Ul> as std::ops::Mul<NInt<Ur>>>
              <PInt<U> as std::ops::Mul<TArr<V, A>>>
note: required by a bound in `si::float::f32::<impl Quantity<D, U, f32>>::powi`
   --> /home/raphael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/uom-0.36.0/src/si/mod.rs:10:1
    |
10  | / system! {
11  | |     /// [International System of Quantities](https://jcgm.bipm.org/vim/en/1.6.html) (ISQ).
12  | |     ///
13  | |     /// ## Generic Parameters
...   |
160 | |     }
161 | | }
    | | ^
    | | |
    | |_required by a bound in this associated function
    |   required by this bound in `si::float::f32::<impl Quantity<D, U, f32>>::powi`
    = note: this error originates in the macro `system` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `{integer}: Integer` is not satisfied
   --> exa/src/main.rs:7:39
    |
7   |     let length: Length = (pos[0].powi(2) + pos[1].powi(2) + pos[2].powi(2)).sqrt();
    |                                  ---- ^ the trait `Integer` is not implemented for `{integer}`
    |                                  |
    |                                  required by a bound introduced by this call
    |
    = help: the following other types implement trait `Integer`:
              PInt<U>
              NInt<U>
              Z0
note: required by a bound in `si::float::f32::<impl Quantity<D, U, f32>>::powi`
   --> /home/raphael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/uom-0.36.0/src/si/mod.rs:10:1
    |
10  | / system! {
11  | |     /// [International System of Quantities](https://jcgm.bipm.org/vim/en/1.6.html) (ISQ).
12  | |     ///
13  | |     /// ## Generic Parameters
...   |
160 | |     }
161 | | }
    | | ^
    | | |
    | |_required by a bound in this associated function
    |   required by this bound in `si::float::f32::<impl Quantity<D, U, f32>>::powi`
    = note: this error originates in the macro `system` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: cannot multiply `PInt<UInt<UTerm, B1>>` by `{integer}`
   --> exa/src/main.rs:7:56
    |
7   |     let length: Length = (pos[0].powi(2) + pos[1].powi(2) + pos[2].powi(2)).sqrt();
    |                                                   ---- ^ no implementation for `PInt<UInt<UTerm, B1>> * {integer}`
    |                                                   |
    |                                                   required by a bound introduced by this call
    |
    = help: the trait `std::ops::Mul<{integer}>` is not implemented for `PInt<UInt<UTerm, B1>>`
    = help: the following other types implement trait `std::ops::Mul<Rhs>`:
              <PInt<U> as std::ops::Mul<Z0>>
              <PInt<U> as std::ops::Mul<ATerm>>
              <PInt<Ul> as std::ops::Mul<PInt<Ur>>>
              <PInt<Ul> as std::ops::Mul<NInt<Ur>>>
              <PInt<U> as std::ops::Mul<TArr<V, A>>>
note: required by a bound in `si::float::f32::<impl Quantity<D, U, f32>>::powi`
   --> /home/raphael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/uom-0.36.0/src/si/mod.rs:10:1
    |
10  | / system! {
11  | |     /// [International System of Quantities](https://jcgm.bipm.org/vim/en/1.6.html) (ISQ).
12  | |     ///
13  | |     /// ## Generic Parameters
...   |
160 | |     }
161 | | }
    | | ^
    | | |
    | |_required by a bound in this associated function
    |   required by this bound in `si::float::f32::<impl Quantity<D, U, f32>>::powi`
    = note: this error originates in the macro `system` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `{integer}: Integer` is not satisfied
   --> exa/src/main.rs:7:56
    |
7   |     let length: Length = (pos[0].powi(2) + pos[1].powi(2) + pos[2].powi(2)).sqrt();
    |                                                   ---- ^ the trait `Integer` is not implemented for `{integer}`
    |                                                   |
    |                                                   required by a bound introduced by this call
    |
    = help: the following other types implement trait `Integer`:
              PInt<U>
              NInt<U>
              Z0
note: required by a bound in `si::float::f32::<impl Quantity<D, U, f32>>::powi`
   --> /home/raphael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/uom-0.36.0/src/si/mod.rs:10:1
    |
10  | / system! {
11  | |     /// [International System of Quantities](https://jcgm.bipm.org/vim/en/1.6.html) (ISQ).
12  | |     ///
13  | |     /// ## Generic Parameters
...   |
160 | |     }
161 | | }
    | | ^
    | | |
    | |_required by a bound in this associated function
    |   required by this bound in `si::float::f32::<impl Quantity<D, U, f32>>::powi`
    = note: this error originates in the macro `system` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: cannot multiply `PInt<UInt<UTerm, B1>>` by `{integer}`
   --> exa/src/main.rs:7:73
    |
7   |     let length: Length = (pos[0].powi(2) + pos[1].powi(2) + pos[2].powi(2)).sqrt();
    |                                                                    ---- ^ no implementation for `PInt<UInt<UTerm, B1>> * {integer}`
    |                                                                    |
    |                                                                    required by a bound introduced by this call
    |
    = help: the trait `std::ops::Mul<{integer}>` is not implemented for `PInt<UInt<UTerm, B1>>`
    = help: the following other types implement trait `std::ops::Mul<Rhs>`:
              <PInt<U> as std::ops::Mul<Z0>>
              <PInt<U> as std::ops::Mul<ATerm>>
              <PInt<Ul> as std::ops::Mul<PInt<Ur>>>
              <PInt<Ul> as std::ops::Mul<NInt<Ur>>>
              <PInt<U> as std::ops::Mul<TArr<V, A>>>
note: required by a bound in `si::float::f32::<impl Quantity<D, U, f32>>::powi`
   --> /home/raphael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/uom-0.36.0/src/si/mod.rs:10:1
    |
10  | / system! {
11  | |     /// [International System of Quantities](https://jcgm.bipm.org/vim/en/1.6.html) (ISQ).
12  | |     ///
13  | |     /// ## Generic Parameters
...   |
160 | |     }
161 | | }
    | | ^
    | | |
    | |_required by a bound in this associated function
    |   required by this bound in `si::float::f32::<impl Quantity<D, U, f32>>::powi`
    = note: this error originates in the macro `system` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `{integer}: Integer` is not satisfied
   --> exa/src/main.rs:7:73
    |
7   |     let length: Length = (pos[0].powi(2) + pos[1].powi(2) + pos[2].powi(2)).sqrt();
    |                                                                    ---- ^ the trait `Integer` is not implemented for `{integer}`
    |                                                                    |
    |                                                                    required by a bound introduced by this call
    |
    = help: the following other types implement trait `Integer`:
              PInt<U>
              NInt<U>
              Z0
note: required by a bound in `si::float::f32::<impl Quantity<D, U, f32>>::powi`
   --> /home/raphael/.cargo/registry/src/index.crates.io-6f17d22bba15001f/uom-0.36.0/src/si/mod.rs:10:1
    |
10  | / system! {
11  | |     /// [International System of Quantities](https://jcgm.bipm.org/vim/en/1.6.html) (ISQ).
12  | |     ///
13  | |     /// ## Generic Parameters
...   |
160 | |     }
161 | | }
    | | ^
    | | |
    | |_required by a bound in this associated function
    |   required by this bound in `si::float::f32::<impl Quantity<D, U, f32>>::powi`
    = note: this error originates in the macro `system` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `exa` (bin "exa" test) due to 6 previous errors
```

Any clarification would be very much welcome!

`uom = "0.36.0"`

**Comment by mkalte666** [comment]
Due to the way uom does what it does under the hood, very specifically pow() is a bit difficult to use - the *compiler* needs to know, at compiletime, what the power is. And since const eval is not that perfect yet, powi() cannot take an int - it cause thats something set at runtime. 

Instead of just putting in an integer, you have to put in a type hint that says "**P**ositive 1, 2, 3" or "**N**egative 1,2,3" etc - and that comes out of the typenum crate. uom then takes the type of that parameter and uses it to construct the type of the result. These are called N2,N1, Z0, P1,P2,P3 (https://docs.rs/typenum/latest/typenum/consts/index.html) 
 
Look here: https://docs.rs/uom/latest/uom/si/struct.Quantity.html#method.powi for an example powi usage. 

In short, your code needs to become this (forgive me if there are typos, i didn't run it locally): 

```rust
use uom::si::f32::*;
use uom::si::length::meter;
use uom::typenum::P2;

fn main() {
    let pos: [Length; 3] = [Length::new::<meter>(1.0), Length::new::<meter>(2.0), Length::new::<meter>(3.0)];
    let length: Length = (pos[0].powi(P2::new()) + pos[1].powi(P2::new()) + pos[2].powi(P2::new())).sqrt();
}
```

Hope this helps!
~Malte

**Comment by RaphaelRobidas** [comment]
Thanks a lot for the precision! Makes sense, it seems like I missed that in the docs.
